man/fsverity.1
.\" SPDX-License-Identifier: MIT
.\"
.\" Copyright 2021 Google LLC
.\"
.\" Use of this source code is governed by an MIT-style license that can be
.\" found in the LICENSE file or at https://opensource.org/licenses/MIT.
.\"
.TH "FSVERITY" "1" "November 2025" "fsverity-utils v1.7" "User Commands"
.hy
.
.SH NAME
.PP
fsverity - userspace utility for fs-verity
.
.
.SH SYNOPSIS
.PP
\f[B]fsverity digest\f[R] [\f[I]OPTION\f[R]\&...]
\f[I]FILE\f[R]\&...
.PD 0
.P
.PD
\f[B]fsverity dump_metadata\f[R] [\f[I]OPTION\f[R]\&...]
\f[I]TYPE\f[R] \f[I]FILE\f[R]
.PD 0
.P
.PD
\f[B]fsverity enable\f[R] [\f[I]OPTION\f[R]\&...]
\f[I]FILE\f[R]
.PD 0
.P
.PD
\f[B]fsverity measure\f[R] \f[I]FILE\f[R]\&...
.PD 0
.P
.PD
\f[B]fsverity sign\f[R] [\f[I]OPTION\f[R]\&...]
\f[I]FILE\f[R] \f[I]OUT_SIGFILE\f[R]
.
.
.SH DESCRIPTION
.PP
\f[B]fsverity\f[R] is a userspace utility for fs-verity.
fs-verity is a Linux kernel filesystem feature that does transparent on-demand
verification of the contents of read-only files using Merkle trees.
.PP
\f[B]fsverity\f[R] can enable fs-verity on files, retrieve the digests of
fs-verity files, and sign files for use with fs-verity (among other things).
\f[B]fsverity\f[R]\[cq]s functionality is divided among various subcommands.
.PP
This manual page focuses on documenting all \f[B]fsverity\f[R] subcommands and
options.
For examples and more information about the fs-verity kernel feature, see the
references at the end of this page.
.
.
.SH OPTIONS
.PP
\f[B]fsverity\f[R] always accepts the following options:
.TP
\f[B]--help\f[R]
Show the help, for either one subcommand or for all subcommands.
.TP
\f[B]--version\f[R]
Show the version of fsverity-utils.
.
.
.SH SUBCOMMANDS
.
.SS \f[B]fsverity digest\f[R] [\f[I]OPTION\f[R]\&...] \f[I]FILE\f[R]\&...
.PP
Compute the fs-verity digest of the given file(s).
This is mainly intended to used in preparation for signing the digest.
In some cases \f[B]fsverity sign\f[R] can be used instead to digest and sign the
file in one step.
.PP
Options accepted by \f[B]fsverity digest\f[R]:
.TP
\f[B]--block-size\f[R]=\f[I]BLOCK_SIZE\f[R]
The Merkle tree block size (in bytes) to use.
This must be a power of 2 and at least twice the size of the hash values.
.RS
.PP
Note that the Linux kernel implementations of fs-verity place further
restrictions on the Merkle tree block size.
Linux v6.2 and earlier require that the Merkle tree block size be equal to both
the system page size and the filesystem block size.
These values are often 4096.
Linux v6.3 and later are more flexible; they require that the Merkle tree block
size be a power of 2 that is greater than or equal to 1024 and less than or
equal to the system page size and the filesystem block size.
The default value of this option is 4096.
.RE
.TP
\f[B]--compact\f[R]
When printing the file digest, only print the actual digest hex string;
don\[cq]t print the algorithm name and filename.
.TP
\f[B]--for-builtin-sig\f[R]
Format the file digest in a way that is compatible with the Linux kernel\[cq]s
fs-verity built-in signature verification support.
This means formatting it as a \f[B]struct fsverity_formatted_digest\f[R].
Use this option if you are using built-in signatures but are not using
\f[B]fsverity sign\f[R] to do the signing.
.TP
\f[B]--hash-alg\f[R]=\f[I]HASH_ALG\f[R]
The hash algorithm to use to build the Merkle tree.
Valid options are sha256 and sha512.
Default is sha256.
.TP
\f[B]--out-merkle-tree\f[R]=\f[I]FILE\f[R]
Write the computed Merkle tree to the given file.
The Merkle tree layout will be the same as that used by the Linux kernel\[cq]s
\f[B]FS_IOC_READ_VERITY_METADATA\f[R] ioctl.
.RS
.PP
Normally this option isn\[cq]t useful, but it can be needed in cases where the
fs-verity metadata needs to be consumed by something other than one of the
native Linux kernel implementations of fs-verity.
This is not needed for file signing.
.RE
.TP
\f[B]--out-descriptor\f[R]=\f[I]FILE\f[R]
Write the computed fs-verity descriptor to the given file.
.RS
.PP
Normally this option isn\[cq]t useful, but it can be needed in cases where the
fs-verity metadata needs to be consumed by something other than one of the
native Linux kernel implementations of fs-verity.
This is not needed for file signing.
.RE
.TP
\f[B]--salt\f[R]=\f[I]SALT\f[R]
The salt to use in the Merkle tree, as a hex string.
The salt is a value that is prepended to every hashed block; it can be used to
personalize the hashing for a particular file or device.
The default is no salt.
.
.SS \f[B]fsverity dump_metadata\f[R] [\f[I]OPTION\f[R]\&...] \f[I]TYPE\f[R] \f[I]FILE\f[R]
.PP
Dump the fs-verity metadata of the given file.
The file must have fs-verity enabled, and the filesystem must support the
\f[B]FS_IOC_READ_VERITY_METADATA\f[R] ioctl (it was added in Linux v5.12).
This subcommand normally isn\[cq]t useful, but it can be useful in cases where a
userspace server program is serving a verity file to a client which implements
fs-verity compatible verification.
.PP
\f[I]TYPE\f[R] may be \[lq]merkle_tree\[rq], \[lq]descriptor\[rq], or
\[lq]signature\[rq], indicating the type of metadata to dump.
\[lq]signature\[rq] refers to the built-in signature, if present;
userspace-managed signatures will not be included.
.PP
Options accepted by \f[B]fsverity dump_metadata\f[R]:
.TP
\f[B]--length\f[R]=\f[I]LENGTH\f[R]
Length in bytes to dump from the specified metadata item.
Only accepted in combination with \f[B]--offset\f[R].
.TP
\f[B]--offset\f[R]=\f[I]offset\f[R]
Offset in bytes into the specified metadata item at which to start dumping.
Only accepted in combination with \f[B]--length\f[R].
.
.SS \f[B]fsverity enable\f[R] [\f[I]OPTION\f[R]\&...] \f[I]FILE\f[R]
.PP
Enable fs-verity on the specified file.
This will only work if the filesystem supports fs-verity.
.PP
Options accepted by \f[B]fsverity enable\f[R]:
.TP
\f[B]--block-size\f[R]=\f[I]BLOCK_SIZE\f[R]
Same as for \f[B]fsverity digest\f[R].
.TP
\f[B]--hash-alg\f[R]=\f[I]HASH_ALG\f[R]
Same as for \f[B]fsverity digest\f[R].
.TP
\f[B]--salt\f[R]=\f[I]SALT\f[R]
Same as for \f[B]fsverity digest\f[R].
.TP
\f[B]--signature\f[R]=\f[I]SIGFILE\f[R]
Specifies the built-in signature to apply to the file.
\f[I]SIGFILE\f[R] must be a file that contains the signature in PKCS#7 DER
format, e.g.\ as produced by the \f[B]fsverity sign\f[R] command.
.RS
.PP
Note that this option is only needed if the Linux kernel\[cq]s fs-verity
built-in signature verification support is being used.
It is not needed if the signatures will be verified in userspace, as in that
case the signatures should be stored separately.
.RE
.
.SS \f[B]fsverity measure\f[R] \f[I]FILE\f[R]\&...
.PP
Display the fs-verity digest of the given file(s).
The files must have fs-verity enabled.
The output will be the same as \f[B]fsverity digest\f[R] with the appropriate
parameters, but \f[B]fsverity measure\f[R] will take constant time for each file
regardless of the size of the file.
.PP
\f[B]fsverity measure\f[R] does not accept any options.
.
.SS \f[B]fsverity sign\f[R] [\f[I]OPTION\f[R]\&...] \f[I]FILE\f[R] \f[I]OUT_SIGFILE\f[R]
.PP
Sign the given file for fs-verity, in a way that is compatible with the Linux
kernel\[cq]s fs-verity built-in signature verification support.
The signature will be written to \f[I]OUT_SIGFILE\f[R] in PKCS#7 DER format.
.PP
The private key can be specified either by key file or by PKCS#11 token.
To use a key file, provide \f[B]--key\f[R] and optionally \f[B]--cert\f[R].
To use a PKCS#11 token, provide \f[B]--pkcs11-engine\f[R],
\f[B]--pkcs11-module\f[R], \f[B]--cert\f[R], and optionally
\f[B]--pkcs11-keyid\f[R].
PKCS#11 token support is unavailable if fsverity-utils was built with a version
of OpenSSL that does not support PKCS#11 tokens.
.PP
\f[B]fsverity sign\f[R] should only be used if you need compatibility with
fs-verity built-in signatures.
It is not the only way to do signatures with fs-verity.
For more information, see the fsverity-utils README.
.PP
Options accepted by \f[B]fsverity sign\f[R]:
.TP
\f[B]--block-size\f[R]=\f[I]BLOCK_SIZE\f[R]
Same as for \f[B]fsverity digest\f[R].
.TP
\f[B]--cert\f[R]=\f[I]CERTFILE\f[R]
Specifies the file that contains the certificate, in PEM format.
This option is required if \f[I]KEYFILE\f[R] contains only the private key and
not also the certificate, or if a PKCS#11 token is used.
.TP
\f[B]--hash-alg\f[R]=\f[I]HASH_ALG\f[R]
Same as for \f[B]fsverity digest\f[R].
.TP
\f[B]--key\f[R]=\f[I]KEYFILE\f[R]
Specifies the file that contains the private key, in PEM format.
This option is required when not using a PKCS#11 token.
.TP
\f[B]--out-descriptor\f[R]=\f[I]FILE\f[R]
Same as for \f[B]fsverity digest\f[R].
.TP
\f[B]--out-merkle-tree\f[R]=\f[I]FILE\f[R]
Same as for \f[B]fsverity digest\f[R].
.TP
\f[B]--pkcs11-engine\f[R]=\f[I]SOFILE\f[R]
Specifies the path to the OpenSSL PKCS#11 engine file.
This typically will be a path to the libp11 .so file.
This option is required when using a PKCS#11 token.
.TP
\f[B]--pkcs11-keyid\f[R]=\f[I]KEYID\f[R]
Specifies the key identifier in the form of a PKCS#11 URI.
If not provided, the default key associated with the token is used.
This option is only applicable when using a PKCS#11 token.
.TP
\f[B]--pkcs11-module\f[R]=\f[I]SOFILE\f[R]
Specifies the path to the PKCS#11 token-specific module library.
This option is required when using a PKCS#11 token.
.TP
\f[B]--salt\f[R]=\f[I]SALT\f[R]
Same as for \f[B]fsverity digest\f[R].
.
.
.SH SEE ALSO
.PP
For example commands and more information, see the README file for
fsverity-utils (https://git.kernel.org/pub/scm/fs/fsverity/fsverity-utils.git/tree/README.md).
.PP
Also see the kernel documentation for
fs-verity (https://www.kernel.org/doc/html/latest/filesystems/fsverity.html).


NEWS.md
# fsverity-utils release notes

## Version 1.7

* Allowed fsverity-utils to build when PKCS#11 support is not available in
  OpenSSL.  In that case, `fsverity sign` just won't support that feature.

## Version 1.6

* Eliminated the dependency on `pandoc` for installing the manual page.

* Updated the documentation to reflect recent kernel changes, including the
  kernel adding support for more Merkle tree block sizes, IMA adding support for
  fs-verity, and btrfs adding support for fs-verity.

* Updated the documentation to refer to the new fsverity mailing list.

* Fixed a C++ compatibility issue in `libfsverity.h`.

* `fsverity measure` now follows standard command line syntax for arguments
  beginning with hyphens.  I.e., `fsverity measure --foo` now treats `--foo` as
  an (unsupported) option, not a filename.  To operate on a file actually named
  `--foo`, use `fsverity measure -- --foo`.

## Version 1.5

* Made the `fsverity sign` command and the `libfsverity_sign_digest()` function
  support PKCS#11 tokens.

* Avoided a compiler error when building with musl libc.

* Avoided compiler warnings when building with OpenSSL 3.0.

* Improved documentation and test scripts.

## Version 1.4

* Added a manual page for the `fsverity` utility.

* Added the `fsverity dump_metadata` subcommand.

* Added the `--out-merkle-tree` and `--out-descriptor` options to
  `fsverity digest` and `fsverity sign`.

* Added metadata callbacks support to `libfsverity_compute_digest()`.

## Version 1.3

* Added a `fsverity digest` subcommand.

* Added `libfsverity_enable()` and `libfsverity_enable_with_sig()`.

* Added basic support for Windows builds of `fsverity` using MinGW.

* `fsverity` now defaults to 4096-byte blocks on all platforms.

* libfsverity now will use SHA-256 with 4096-byte blocks if the
  `hash_algorithm` and `block_size` fields are left 0.

* `make install` now installs a pkg-config file for libfsverity.

* The Makefile now uses pkg-config to get the libcrypto build flags.

* Fixed `make check` with `USE_SHARED_LIB=1`.

## Version 1.2

* Changed license from GPL to MIT.

* Fixed build error when /bin/sh is dash.

## Version 1.1

* Split the file digest computation and signing functionality of the
  `fsverity` program into a library `libfsverity`.  See `README.md`
  and `Makefile` for more details.

* Improved the Makefile.

* Added some tests.  They can be run using `make check`.  Also added
  `scripts/run-tests.sh` which does more extensive prerelease tests.

* Lots of cleanups and other small improvements.

## Version 1.0

* First official release of fsverity-utils.


programs/fsverity.c
// SPDX-License-Identifier: MIT
/*
 * fs-verity userspace tool
 *
 * Copyright 2018 Google LLC
 *
 * Use of this source code is governed by an MIT-style
 * license that can be found in the LICENSE file or at
 * https://opensource.org/licenses/MIT.
 */

#include "fsverity.h"

#include <fcntl.h>
#include <limits.h>

static const struct fsverity_command {
	const char *name;
	int (*func)(const struct fsverity_command *cmd, int argc, char *argv[]);
	const char *short_desc;
	const char *usage_str;
} fsverity_commands[] = {
	{
		.name = "digest",
		.func = fsverity_cmd_digest,
		.short_desc =
"Compute the fs-verity digest of the given file(s), for offline signing",
		.usage_str =
"    fsverity digest FILE...\n"
"               [--hash-alg=HASH_ALG] [--block-size=BLOCK_SIZE] [--salt=SALT]\n"
"               [--out-merkle-tree=FILE] [--out-descriptor=FILE]\n"
"               [--compact] [--for-builtin-sig]\n"
#ifndef _WIN32
	}, {
		.name = "dump_metadata",
		.func = fsverity_cmd_dump_metadata,
		.short_desc = "Dump the fs-verity metadata of the given file",
		.usage_str =
"    fsverity dump_metadata TYPE FILE [--offset=OFFSET] [--length=LENGTH]\n"
	}, {
		.name = "enable",
		.func = fsverity_cmd_enable,
		.short_desc = "Enable fs-verity on a file",
		.usage_str =
"    fsverity enable FILE\n"
"               [--hash-alg=HASH_ALG] [--block-size=BLOCK_SIZE] [--salt=SALT]\n"
"               [--signature=SIGFILE]\n"
	}, {
		.name = "measure",
		.func = fsverity_cmd_measure,
		.short_desc =
"Display the fs-verity digest of the given verity file(s)",
		.usage_str =
"    fsverity measure FILE...\n"
#endif /* !_WIN32 */
	}, {
		.name = "sign",
		.func = fsverity_cmd_sign,
		.short_desc = "Sign a file for fs-verity",
		.usage_str =
"    fsverity sign FILE OUT_SIGFILE\n"
"               [--key=KEYFILE] [--cert=CERTFILE] [--pkcs11-engine=SOFILE]\n"
"               [--pkcs11-module=SOFILE] [--pkcs11-keyid=KEYID]\n"
"               [--hash-alg=HASH_ALG] [--block-size=BLOCK_SIZE] [--salt=SALT]\n"
"               [--out-merkle-tree=FILE] [--out-descriptor=FILE]\n"
	}
};

static void show_all_hash_algs(FILE *fp)
{
	u32 alg_num = 1;
	const char *name;

	fprintf(fp, "Available hash algorithms:");
	while ((name = libfsverity_get_hash_name(alg_num++)) != NULL)
		fprintf(fp, " %s", name);
	putc('\n', fp);
}

static void usage_all(FILE *fp)
{
	int i;

	fputs("Usage:\n", fp);
	for (i = 0; i < ARRAY_SIZE(fsverity_commands); i++)
		fprintf(fp, "  %s:\n%s\n", fsverity_commands[i].short_desc,
			fsverity_commands[i].usage_str);
	fputs(
"  Standard options:\n"
"    fsverity --help\n"
"    fsverity --version\n"
"\n", fp);
	show_all_hash_algs(fp);
}

static void usage_cmd(const struct fsverity_command *cmd, FILE *fp)
{
	fprintf(fp, "Usage:\n%s", cmd->usage_str);
}

void usage(const struct fsverity_command *cmd, FILE *fp)
{
	if (cmd)
		usage_cmd(cmd, fp);
	else
		usage_all(fp);
}

static void show_version(void)
{
	printf("fsverity v%d.%d\n", FSVERITY_UTILS_MAJOR_VERSION,
	       FSVERITY_UTILS_MINOR_VERSION);
}

static void handle_common_options(int argc, char *argv[],
				  const struct fsverity_command *cmd)
{
	int i;

	for (i = 1; i < argc; i++) {
		const char *arg = argv[i];

		if (*arg++ != '-')
			continue;
		if (*arg++ != '-')
			continue;
		if (!strcmp(arg, "help")) {
			usage(cmd, stdout);
			exit(0);
		} else if (!strcmp(arg, "version")) {
			show_version();
			exit(0);
		} else if (!*arg) /* reached "--", no more options */
			return;
	}
}

static const struct fsverity_command *find_command(const char *name)
{
	int i;

	for (i = 0; i < ARRAY_SIZE(fsverity_commands); i++)
		if (!strcmp(name, fsverity_commands[i].name))
			return &fsverity_commands[i];
	return NULL;
}

static bool parse_hash_alg_option(const char *arg, u32 *alg_ptr)
{
	char *end;
	unsigned long n = strtoul(arg, &end, 10);

	if (*alg_ptr != 0) {
		error_msg("--hash-alg can only be specified once");
		return false;
	}

	/* Specified by number? */
	if (n > 0 && n < INT32_MAX && *end == '\0') {
		*alg_ptr = n;
		return true;
	}

	/* Specified by name? */
	*alg_ptr = libfsverity_find_hash_alg_by_name(arg);
	if (*alg_ptr)
		return true;
	error_msg("unknown hash algorithm: '%s'", arg);
	show_all_hash_algs(stderr);
	return false;
}

static bool parse_block_size_option(const char *arg, u32 *size_ptr)
{
	char *end;
	unsigned long n = strtoul(arg, &end, 10);

	if (*size_ptr != 0) {
		error_msg("--block-size can only be specified once");
		return false;
	}

	if (n <= 0 || n >= INT_MAX || !is_power_of_2(n) || *end != '\0') {
		error_msg("Invalid block size: %s.  Must be power of 2", arg);
		return false;
	}
	*size_ptr = n;
	return true;
}

static bool parse_salt_option(const char *arg, u8 **salt_ptr,
			      u32 *salt_size_ptr)
{
	if (*salt_ptr != NULL) {
		error_msg("--salt can only be specified once");
		return false;
	}
	*salt_size_ptr = strlen(arg) / 2;
	*salt_ptr = xmalloc(*salt_size_ptr);
	if (!hex2bin(arg, *salt_ptr, *salt_size_ptr)) {
		error_msg("salt is not a valid hex string");
		return false;
	}
	return true;
}

struct metadata_callback_ctx {
	struct filedes merkle_tree_file;
	struct filedes descriptor_file;
	struct libfsverity_metadata_callbacks callbacks;
};

static int handle_merkle_tree_size(void *_ctx, u64 size)
{
	struct metadata_callback_ctx *ctx = _ctx;

	if (!preallocate_file(&ctx->merkle_tree_file, size))
		return -EIO;
	return 0;
}

static int handle_merkle_tree_block(void *_ctx, const void *block, size_t size,
				    u64 offset)
{
	struct metadata_callback_ctx *ctx = _ctx;

	if (!full_pwrite(&ctx->merkle_tree_file, block, size, offset))
		return -EIO;
	return 0;
}

static int handle_descriptor(void *_ctx, const void *descriptor, size_t size)
{
	struct metadata_callback_ctx *ctx = _ctx;

	if (!full_write(&ctx->descriptor_file, descriptor, size))
		return -EIO;
	return 0;
}

static bool parse_out_metadata_option(int opt_char, const char *arg,
				      const struct libfsverity_metadata_callbacks **cbs)
{
	struct metadata_callback_ctx *ctx;
	struct filedes *file;
	const char *opt_name;

	if (*cbs) {
		ctx = (*cbs)->ctx;
	} else {
		ctx = xzalloc(sizeof(*ctx));
		ctx->merkle_tree_file.fd = -1;
		ctx->descriptor_file.fd = -1;
		ctx->callbacks.ctx = ctx;
		*cbs = &ctx->callbacks;
	}

	if (opt_char == OPT_OUT_MERKLE_TREE) {
		file = &ctx->merkle_tree_file;
		opt_name = "--out-merkle-tree";
		ctx->callbacks.merkle_tree_size = handle_merkle_tree_size;
		ctx->callbacks.merkle_tree_block = handle_merkle_tree_block;
	} else {
		file = &ctx->descriptor_file;
		opt_name = "--out-descriptor";
		ctx->callbacks.descriptor = handle_descriptor;
	}
	if (file->fd >= 0) {
		error_msg("%s can only be specified once", opt_name);
		return false;
	}
	return open_file(file, arg, O_WRONLY|O_CREAT|O_TRUNC, 0644);
}

bool parse_tree_param(int opt_char, const char *arg,
		      struct libfsverity_merkle_tree_params *params)
{
	switch (opt_char) {
	case OPT_HASH_ALG:
		return parse_hash_alg_option(arg, &params->hash_algorithm);
	case OPT_BLOCK_SIZE:
		return parse_block_size_option(arg, &params->block_size);
	case OPT_SALT:
		return parse_salt_option(arg, (u8 **)&params->salt,
					 &params->salt_size);
	case OPT_OUT_MERKLE_TREE:
	case OPT_OUT_DESCRIPTOR:
		return parse_out_metadata_option(opt_char, arg,
						 &params->metadata_callbacks);
	default:
		ASSERT(0);
	}
}

bool destroy_tree_params(struct libfsverity_merkle_tree_params *params)
{
	bool ok = true;

	free((u8 *)params->salt);
	if (params->metadata_callbacks) {
		struct metadata_callback_ctx *ctx =
			params->metadata_callbacks->ctx;

		ok &= filedes_close(&ctx->merkle_tree_file);
		ok &= filedes_close(&ctx->descriptor_file);
		free(ctx);
	}
	memset(params, 0, sizeof(*params));
	return ok;
}

int main(int argc, char *argv[])
{
	const struct fsverity_command *cmd;

	install_libfsverity_error_handler();

	if (argc < 2) {
		error_msg("no command specified");
		usage_all(stderr);
		return 2;
	}

	cmd = find_command(argv[1]);

	handle_common_options(argc, argv, cmd);

	if (!cmd) {
		error_msg("unrecognized command: '%s'", argv[1]);
		usage_all(stderr);
		return 2;
	}
	return cmd->func(cmd, argc - 1, argv + 1);
}
